pymc_marketing/clv/distributions.py

```diff
@@ -19,59 +19,47 @@ class ContNonContractRV(RandomVariable):
     dtype = "floatX"
     _print_name = ("ContNonContract", "\\operatorname{ContNonContract}")
 
-    def make_node(self, rng, size, dtype, lam, p, T, T0):
+    def make_node(self, rng, size, dtype, lam, p, T):
         T = pt.as_tensor_variable(T)
-        T0 = pt.as_tensor_variable(T0)
 
-        return super().make_node(rng, size, dtype, lam, p, T, T0)
-
-    def __call__(self, lam, p, T, T0=0, size=None, **kwargs):
-        return super().__call__(lam, p, T, T0, size=size, **kwargs)
+        return super().make_node(rng, size, dtype, lam, p, T)
 
     @classmethod
-    def rng_fn(cls, rng, lam, p, T, T0, size) -> np.ndarray:
+    def rng_fn(cls, rng, lam, p, T, size) -> np.ndarray:
         size = pm.distributions.shape_utils.to_tuple(size)
 
         # TODO: broadcast sizes
         lam = np.asarray(lam)
         p = np.asarray(p)
         T = np.asarray(T)
-        T0 = np.asarray(T0)
 
         if size == ():
-            size = np.broadcast_shapes(lam.shape, p.shape, T.shape, T0.shape)
+            size = np.broadcast_shapes(lam.shape, p.shape, T.shape)
 
         lam = np.broadcast_to(lam, size)
         p = np.broadcast_to(p, size)
         T = np.broadcast_to(T, size)
-        T0 = np.broadcast_to(T0, size)
 
-        output = np.zeros(shape=size + (2,))
+        x_1 = rng.poisson(lam * T)
+        x_2 = rng.geometric(p)
+        x = np.minimum(x_1, x_2)
 
-        # TODO: Optimize to work in a vectorized manner!
-        def sim_data(lam, p, T, T0):
-            t = T0
-            n = 0
-
-            while True:
-                wait = rng.exponential(scale=1 / lam)
-                dropout = rng.binomial(n=1, p=p)
-
-                if t + wait > T:
-                    break
-                else:
-                    t += wait
-                    n += 1
+        def modified_rng_beta(a, b):
+            if a == 0:
+                "If N(T) = 0, then t_x = t_0 = 0."
+                return 0
 
-                    if dropout == 1:
-                        break
+            return rng.beta(a, b)
 
-            return np.array([t, n])
-
-        for index in np.ndindex(*size):
-            output[index] = sim_data(lam[index], p[index], T[index], T0[index])
+        t_x = (
+            np.vectorize(modified_rng_beta)(
```

Comment:
This will force a slow loop through `(a, b)` pairs. A better approach would be to call `beta` on the inputs and then override with zeros where needed:

```python
t_x = rng.beta(x, np.maximum(...))
t_x[x == 0] = 0
```

And suppress any warnings if needed

Reply:
`rng.beta(a, b)` throws an error for `a <= 0`. Perhaps the following would work better?
```python
t_x[x > 0] = rng.beta(x, np.maximum(...))
```

Reply:
Does NumPy some sort of lazy evaluation where `rng.beta(x, ...)` does not need to be evaluated where `x == 0`?

Reply:
Not sure. You can index yourself. Are x=0 common?

```python
ret = np.zeros()
ret[x~0] = rng.beta(a[x~0], b[x~0])
```

But if they are not common I wouldn't bother [the two extra slices could be slower than the useless beta). Unless numpy raises ofc

Reply:
Depending on the hazard parameter, it is generally uncommon, but important to account for. It characterizes the situation where there is too little time to observe if any purchases can be made.

I'll try what you suggest above later this afternoon

Reply:
I meant I wouldn't bother to index before and would just override with zeros after (if zeros don't cause it to crash)

Reply:
Zeros still makes this crash... with `size` is just 1D, I can use
```python
nzp = x != 0 # nzp = non-zero purchases

        t_x = np.zeros(shape=size)
        t_x[nzp] = rng.beta(x[nzp], np.maximum(x_1 + 1 - x_2, 1)[nzp])*T
```
but this does not work when `size` is 2D (e.g. `(5, 3)` in tests). I'm not sure how to mitigate this issue...